### PR TITLE
Add sources and refactor `nameable` to support polymorphic shared name tables

### PR DESCRIPTION
**Pull request by camdendotlol**
# Summary

Accompanies https://github.com/performant-software/core-data-cloud/pull/58 on the frontend.

- adds three new `Ownable`, `Relatable`, and UDF-enabled models: `Work`, `Item`, and `Instance`
- refactors `nameable` to allow a `polymorphic` boolean configuration option that shares a polymorphic M2M table between multiple models
  - enables this for the three source models, to share their titles via `source_titles`
- adds a new `Name` model and associated logic where polymorphic `nameable` names will be stored
- updates the Search concern for the new models

### CODE REVIEW

File: app/models/core_data_connector/title_source.rb
Comment (dleadbetter) [inline app/models/core_data_connector/title_source.rb:1]: Interesting. Do we anticipate titles being reused? Why not related the `title` directly to the `work` or `instance`?

**Review by dleadbetter — COMMENTED** [review]
This looks like a great start. I would recommend implementing the front-end as part of this task as well before an official review. I think there are a few things missing that will come to light.

This project is also a little different from our typical projects that use this stack in that it may at some point be used as a SaaS platform. The big difference there is that a user's visibility to data should be scoped to the projects of which they are a member. The `resource-api` now has built in support for [Pundit](https://github.com/varvet/pundit), which we should use to create policies for any models that will be accessed via the API to ensure correct authorization.

Your point about the `items`, `instances`, and `works` models being redundant is valid. @jamiefolsom may have some thoughts on this, but it's my understanding that these models are essentially stubs at this point. As we implement more projects that are heavily focused on sources, they'll help us enrich these models with more native fields.

**Comment by jamiefolsom** [comment]
> This looks like a great start. I would recommend implementing the front-end as part of this task as well before an official review. I think there are a few things missing that will come to light.
> 
> This project is also a little different from our typical projects that use this stack in that it may at some point be used as a SaaS platform. The big difference there is that a user's visibility to data should be scoped to the projects of which they are a member. The `resource-api` now has built in support for [Pundit](https://github.com/varvet/pundit), which we should use to create policies for any models that will be accessed via the API to ensure correct authorization.
> 
> Your point about the `items`, `instances`, and `works` models being redundant is valid. @jamiefolsom may have some thoughts on this, but it's my understanding that these models are essentially stubs at this point. As we implement more projects that are heavily focused on sources, they'll help us enrich these models with more native fields.

There are semantic differences between items and instances and works, and I suspect there may be native fields requested by projects which would be make sense to include directly on these tables, but which we may not need them now.

An alternative, that occurs to me would be to use [single table inheritance](https://guides.rubyonrails.org/association_basics.html#single-table-inheritance-sti), where the primary model/table is called source, and it has the superset of all columns necessary for works and instances and editions. Currently there are none, but that might change over time. Then we could have a single parent `Source` model in the Rails application, and child models for `Work`, `Instance` and `Edition`, each of which would implement support for the specific columns they need.

The table might be somewhat sparse, but maybe that's OK. Thoughts on that approach, Derek?

File: app/models/core_data_connector/title_source.rb
Comment (camdendotlol) [reply app/models/core_data_connector/title_source.rb:1]: Should titles just be another model that users can add to their project?

Comment (dleadbetter) [reply app/models/core_data_connector/title_source.rb:1]: I think the titles for a source would be analogous to how we're allowing multiple names for a place/person/organization.

File: config/routes.rb
Comment (dleadbetter) [inline config/routes.rb:23]: For consistency, let's set this up as a `resources` statement as well, only allowing the `index` route.

File: db/migrate/20231115163827_create_core_data_connector_works.rb
Comment (dleadbetter) [inline db/migrate/20231115163827_create_core_data_connector_works.rb:5]: Let's use the generator in the `user_defined_fields` gem to add the columns to the table. It also creates a database index to help performance.

File: lib/typesense/helper.rb
Comment (dleadbetter) [inline lib/typesense/helper.rb:100]: This can probably be reverted?

**Review by dleadbetter — COMMENTED** [review]
I would also take a look into the `relationships_controller`. There may be some work to do there to allow searching/sorting of related items/instances/works. For example, if you're in the context of a record and looking at all of the related `works`, are you able to filter that list by searching for a name?

File: app/controllers/core_data_connector/works_controller.rb
Comment (camdendotlol) [reply app/controllers/core_data_connector/works_controller.rb:7]: So the only place this `name_class` is accessed is in the `resolve_search_attribute` method in `NameableController`. I can't figure out where that method is called - nothing else in the codebase directly calls it and when I add `puts 'hello'` I don't see anything when either searching in the Rails app UI or when indexing to Typesense.

If `resolve_search_attribute` is actually dead code, we can just remove the `name_class` thing completely, but I assume it's being used by something?

Comment (camdendotlol) [reply app/controllers/core_data_connector/works_controller.rb:7]: Ah, never mind. I see it now in the Rails app search.

File: app/controllers/core_data_connector/names_controller.rb
Comment (camdendotlol) [reply app/controllers/core_data_connector/names_controller.rb:14]: I think I see what you're saying - the `names` table isn't `JOIN`ed in the second `WHERE` clause?

Does this work? It returns the correct results when I test it.

```ruby
        .where(
          nameable_model
            .joins(nameable_model.get_name_table.to_sym => :name) # Added this line
            .joins(:project_model)
            .where(
              CoreDataConnector::ProjectModel.arel_table[:project_id].eq(params[:project_id])
            )
            .arel
            .exists
        )
```

File: app/models/concerns/core_data_connector/nameable.rb
Comment (camdendotlol) [reply app/models/concerns/core_data_connector/nameable.rb:5]: So far, on People and Places, I've tested:

1. Creating/editing records with names
2. Sorting and searching by names (this was indeed broken for existing models, and I fixed it on Friday)
3. Proper selection of `primary_name`s

Is there any other name-related functionality that I should check for regressions?

File: app/models/concerns/core_data_connector/nameable.rb
Comment (camdendotlol) [reply app/models/concerns/core_data_connector/nameable.rb]: I spent some time trying to get `mattr_accessor` working (`cattr_accessor` has been removed and is now an alias for `mattr_accessor`) and I eventually got something that didn't error on me based on the examples from https://api.rubyonrails.org/v7.0.5/classes/Module.html#method-i-mattr_accessor.

However, I found two issues with this approach:

1. I had to call `model.class_variable_get("@@names_table")` every time I accessed it, which was more tedious than just `get_name_table`.
2. `mattr_accessor` only seemed to work if I set `names_table` as a class variable (`@@`), and it seems like setting a class variable from a concern sets it for all models that include the concern. In other words, People/Places functionality broke because the value was being set to `source_titles` by the source models.

If there's a way to get `mattr_accessor` to work on a properly scoped, per-model basis, I'm willing to try it, but I feel like I've burned too much time on a minor thing that might not actually come out simpler than `get_name_table` in the end.

File: app/controllers/core_data_connector/names_controller.rb
Comment (camdendotlol) [reply app/controllers/core_data_connector/names_controller.rb:14]:
```sql
SELECT "core_data_connector_names".* FROM "core_data_connector_names" WHERE EXISTS (SELECT "core_data_connector_source_titles".* FROM "core_data_connector_source_titles" WHERE "core_data_connector_source_titles"."name_id" = "core_data_connector_names"."id") AND EXISTS (SELECT "core_data_connector_instances".* FROM "core_data_connector_instances" INNER JOIN "core_data_connector_source_titles" ON "core_data_connector_source_titles"."nameable_type" = 'CoreDataConnector::Instance' AND "core_data_con
2:09:50 PM api.1 |  >  nector_source_titles"."nameable_id" = "core_data_connector_instances"."id" INNER JOIN "core_data_connector_names" ON "core_data_connector_names"."id" = "core_data_connector_source_titles"."name_id" INNER JOIN "core_data_connector_project_models" ON "core_data_connector_project_models"."id" = "core_data_connector_instances"."project_model_id" WHERE "core_data_connector_project_models"."project_id" = 1)
```

Comment (dleadbetter) [reply app/controllers/core_data_connector/names_controller.rb:14]:
```
SELECT "core_data_connector_names".* 
  FROM "core_data_connector_names" 
 WHERE EXISTS ( SELECT "core_data_connector_source_titles".* 
                  FROM "core_data_connector_source_titles" 
                 WHERE "core_data_connector_source_titles"."name_id" = "core_data_connector_names"."id")
   AND EXISTS ( SELECT "core_data_connector_instances".* 
                  FROM "core_data_connector_instances" 
                 INNER JOIN "core_data_connector_source_titles" ON "core_data_connector_source_titles"."nameable_type" = 'CoreDataConnector::Instance' 
                                                               AND "core_data_connector_source_titles"."nameable_id" = "core_data_connector_instances"."id" 
                 INNER JOIN "core_data_connector_names" ON "core_data_connector_names"."id" = "core_data_connector_source_titles"."name_id" 
                 INNER JOIN "core_data_connector_project_models" ON "core_data_connector_project_models"."id" = "core_data_connector_instances"."project_model_id" 
                 WHERE "core_data_connector_project_models"."project_id" = 1)
```

The problem I see is that the second `EXISTS` doesn't filter the outer query at all. It's `JOIN`ing it's own `core_data_connector_names`, which I don't think is what we want to do.

File: app/controllers/core_data_connector/names_controller.rb
Comment (dleadbetter) [inline app/controllers/core_data_connector/names_controller.rb:12]: So we're allow a user to select from a list of names for a particular model within a project. So if I have a `Work` record with a name "A Tale of Two Cities", I can only reuse that name on other `Work` records?

Comment (camdendotlol) [reply app/controllers/core_data_connector/names_controller.rb:12]: The scoping is based on the `polymorphic_table` (a few lines down), in this case `source_titles`. Models that share a polymorphic M2M name relation table should be able to share their list of names. In other words, Instance, Item, and Work should be able to reuse names across all three models because they share the `source_titles` M2M table. Do you see anything here that would make it work differently?

Comment (dleadbetter) [reply app/controllers/core_data_connector/names_controller.rb:12]: So presumably this controller would get a request as:
```
/core_data/names?project_id=1&nameable_model=work
```

Removing all of the meta-programming, the resulting query would look something like this:
```
query
  .where(
    SourceTitle
       .joins(works :project_model)
       .where(project_models: {
         project_id: params[:project_id]
       })
       .where(SourceTitle.arel_table[:nameable_type].eq(Work.to_s))
       .where(SourceTitle.arel_table[:name_id].eq(CoreDataConnector::Name.arel_table[:id]))
       .arel
       .exists
  )
```

Wouldn't the `where(SourceTitle.arel_table[:nameable_type].eq(Work.to_s))` filter this query to only include `source_titles` records that are linked to `works`?

This should also be easily testable by adding a name to one type of model and seeing if it is available when creating another model.

File: app/controllers/core_data_connector/names_controller.rb
Comment (dleadbetter) [inline app/controllers/core_data_connector/names_controller.rb:20]: Since we're in the `CoreDataConnector` module, no need to prefix the class here.

File: app/controllers/core_data_connector/names_controller.rb
Comment (dleadbetter) [inline app/controllers/core_data_connector/names_controller.rb:23]: Ditto above. We can probably keep this to one line as well.

File: app/controllers/core_data_connector/names_controller.rb
Comment (dleadbetter) [inline app/controllers/core_data_connector/names_controller.rb:29]: I like this approach to building the queries since they're so similar. Can we eliminate the need to hard-code array indices? Something like `queries.map(&:to_sql).join(' UNION ')`?

File: app/models/core_data_connector/name.rb
Comment (dleadbetter) [inline app/models/core_data_connector/name.rb:3]: Can you explain a little about what is going on here? I didn't think multiple `has_many` relationships with the same name could be valid.

Comment (camdendotlol) [reply app/models/core_data_connector/name.rb]: Just some dead code from a very early version of the feature - removed.

File: app/controllers/core_data_connector/names_controller.rb
Comment (dleadbetter) [reply app/controllers/core_data_connector/names_controller.rb:27]: 👍 Hopefully this was re-tested after refactoring?

**Review by dleadbetter — APPROVED** [review]
Looks good 👍 . Let's wait until the `core-data-cloud` changes are approved to merge.

Edit: Are we missing some changes to the `relationships_controller` to allow searching/sorting lists of related items/instances/works?

**Comment by camdendotlol** [comment]
> Edit: Are we missing some changes to the `relationships_controller` to allow searching/sorting lists of related items/instances/works?

Done

**Pull request merged**
